### ISSUE TITLE
Look into pxrmt speed

### ISSUE THREAD

**Issue by josephwb** [FePhyFoFum/phyx#74]
The new R package [castor](https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/btx701/4582279) makes some phyx comparisons.

Apparently for a tree of 500,000 tips, removing half the tips takes over 3 hours (at which point the author just cancelled the process).

`pxtrt` probably does this well. I've never tried it. But if the number of tips to prune is large we might use the `pxtrt` functions instead. 

Should test this out on the SILVA tree used in the castor pub.

**Comment by blackrim** [comment]
Do we have a dataset for testing this? if so, we can do `pxtrt`

**Comment by josephwb** [comment]
The [revamped](https://github.com/FePhyFoFum/phyx/issues/114) `pxtrt` program accomplishes the castor 576542-tip example above in 5 seconds (I believe faster than any of the benchmarks in that paper?). 

I am making a new general subtree class that will accomplish all that `pxrmt` (prune tips), `pxtrt` (induced subtree), and `pxmrcacut` (subtree) currently do.

**Comment by josephwb** [comment]
I've been looking into the relative speeds of `pxrmt` and `pxtrt`. Below are some benchmarks across tree sizes (each point is the average across 3 distinct trees):
![results](https://user-images.githubusercontent.com/808499/70817058-bb24a280-1d9e-11ea-882f-74306d2096f1.png)
In general, `pxrmt` is _always_ faster (regardless of tree size) if the number to be removed in small. It is only when the number to be removed is large (say, >= 50) that the overhead involved makes things slower, and indeed this overhead is (not surprisingly) linear with the number of terminal to remove.

I believe that `pxtrt` is just about as efficient as it can get. I have not poked around `pxrmt` much, but I feel like the memory management overhead will be unavoidable.

For the time being, I might just (unsatisfyingly) use some magic numbers to pick which algorithm to use.

**Comment by josephwb** [comment]
As currently implemented, `pxtrt` requires a rooted tree (as traces use the root as a termination). Because of the huge overhead for `pxrmt` on large trees (above), I tried the following for unrooted trees:
1. root on an arbitrary terminal
2. do the trace (i.e., remove specified terminals)
3. unroot

it turns out the added procedures (rooting + unrooting) is negligible (testing a single tree of 50K tips, removing 5K; the worst-case scenario above):
```
pxrmt rooted ntips 500000 remove 5000
414.075

pxtrt rooted ntips 500000 remove 5000
16.342

pxrmt unrooted ntips 500000 remove 5000
412.380

pxtrt unrooted ntips 500000 remove 5000
tree is not rooted.
16.376
```
So it looks like this is a viable approach. To be sure, I'll re-run the analysis illustrated above, but with unrooted trees.

**Comment by josephwb** [comment]
It appears that the approach above is indeed viable. Here is the plot above but for unrooted trees:
![results](https://user-images.githubusercontent.com/808499/70864243-51330700-1f1e-11ea-8997-26cb0dc0c749.png)

**Comment by josephwb** [comment]
I implemented the procedure above with 35a99eab2046258ba3b5d078552554811298ddef. Here are the results for the most extreme case above (500k tips, remove 5 K):
```
# rooted input
Original pxrmt
400.328
New pxrmt
18.014

# unrooted input
Original pxrmt
400.042
New pxrmt
18.411
```
So this looks resolved. I'll close this when merging to master.
